### DISCUSSION TITLE
[Feature] Possibility to disable transcoding of HDR videos

### DISCUSSION THREAD

**Discussion by Hely0n** [immich-app/immich#12844] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

As far as I know, the forced transcode of HDR videos is to ensure compatiblity with browsers, but it won't affect users that use chromium-based browsers for playback. If my understanding is true, I would like to disable the transcoding of HDR videos, because it effectively transcodes all of our iPhone videos unnecessarily, especially when the new native-player will be implemented on mobile.

Am I missing something? If not, I would like to implement a little "transcode HDR videos" switch in the settings that will be enabled by default 😊.

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by mertalev** [comment]
We can revisit this after the mobile app has proper HDR support.

**Reply by Hely0n** [reply]
Thank you!

**Reply by Hely0n** [reply]
Sooo... 😂

BTW, Thank you (and the others of course) for implementing the native player. 
I followed the PR, as this was the last thing that was needed to not having to use the iOS photos app. Now each and everything can be watched inside the Immich app, from yesterdays HDR videos, to photos from the childhood.

**Reply by himankpathak** [reply]
Now that the mobile app has HDR support, can this feature be reconsidered? @mertalev

**Reply by jonathan1107** [reply]
Yes please I want hdr videos to NOT transcode and be fully supported in all the immich app and web players

**Comment by Hely0n** [comment]
I have disabled trancoding and made a short HEVC HDR video with my iPhone. I shared this video by link and to another user and tested it:

It works on iOS (Safari and App), Android (Chrome), Windows (Brave/Chrome). I think the only problem are firefox users, but they coudn't play my transcoded videos either, as they are HEVC and my friends are all using chromium based browsers.

That means, that I lose (currently) ~500GB~ like 150GB space for no reason, and have 100% CPU usage for an hour after me or my girlfriend come home from a trip because all Videos will be transcoded. And what's even worse in my opinion, is the quality downgrade on all videos, when watching them on a TV with others (at least until original playback will be implemented).

I want to start implementing the switch myself now. I hope it's okay 😅

**Reply by Hely0n** [reply]
Android (App) works too

**Reply by maddog107** [reply]
@mertalev 
Since https://github.com/immich-app/immich/issues/5120 is closed (support for HDR), will the option to disable HDR transcoding be available now? I have 90k vidoes that need transcoding because all of them were shot in HDR on the iphone. Thanks!

**Comment by b4stOss** [comment]
Hi,
Is there any news on this feature?
I just spent two hours trying to figure out why my entire iPhone video library is being transcoded, and I found out it’s because HDR overrides the transcoding policy.
It doesn’t make sense to me to waste so much space and energy just to transcode HDR videos while the Immich app and most browsers already support it.
Or am I missing something?

**Reply by Hely0n** [reply]
I gave up on that, because I found out that iPhone videos have a rather high Bitrate and I couldn‘t stream those videos remotely neither via WireGuard nor Cloudflare, despite having 200mbit/s Upload speed. That‘s why I changed to the max bitrate policy , so everything will be transcoded with more than 40M.

So Iphone Videos must be transcoded anyway, independent from being HDR. 🫤

Edit: I also found out that the culprit here isn‘t my upload speed but rather the download speed at many locations from where me or my friends tried to stream it

**Reply by ipnetxyz** [reply]
I have the same issue, was trying to understand why is getting transcoded, even though all settings were as supposed to be...found out that the videos were taken with HDR on.

Any chance this is still considered? Thank you!

**Comment by ipnetxyz** [comment]
### Disable HDR & 10-bit forced transcoding

**Disclaimer:** 
This is a brutal method (like carving a toothpick with an ax) and I used Claude to help with some blocking points.
It requires a bit of skills to deal with docker, etc...but if you're reading this I assume no problem.
The changes are not persistent on the immich stack restart - but I have a workaround a the end of this comment.

Please do not hold me accountable if you break your Immich installation.

**My environment:**
I tested this on the current Immich v2 as of today (May 20th, 2026).
Installation of Immich is done using the official Docker Compose guide on Debian 13.
Videos are 90% made with iPhone using HDR.
The "Transcode Policy" is set to "Only videos not in the accepted format" and have all options checked under Accepted video codecs / Accepted audio codecs / Accepted containers.

**My wish**
Maybe somebody more skilled than me can take the "hack" and make it permanent.

**Patch 1 — Disable HDR as a transcode trigger (utils/media.js):**

```
docker exec immich_server sed -i \
  's/return videoStream\.isHDR && this\.config\.tonemap !== enum_1\.ToneMapping\.Disabled;/return false;/' \
  /usr/src/app/server/dist/utils/media.js
```
This makes shouldToneMap() always return false, so HDR metadata is never used as a reason to transcode.


**Patch 2 — Accept 10-bit pixel formats (services/media.service.js):**

```
docker exec immich_server sed -i \
  "s/stream\.pixelFormat\.endsWith('420p')/stream.pixelFormat.includes('420p')/" \
  /usr/src/app/server/dist/services/media.service.js
```
  
This makes yuv420p10le, yuv420p12le etc. pass the pixel format check instead of triggering a transcode. The code currently looks for pixel format ending explicit with 420p, so anything else (in my case was yuv420p10le) will be transcoded.

After applying both patches, restart:

```
docker restart immich_server
```

Verify both patches:

```
docker exec immich_server grep -n "shouldToneMap\|return false" /usr/src/app/server/dist/utils/media.js | head -5
docker exec immich_server grep -n "pixelFormat" /usr/src/app/server/dist/services/media.service.js
```

To verify functionality, hit Refresh encoded video on an HDR asset. 
You should see in the logs a delete if the asset is HDR and has a transcoded file, or nothing if the is video is HDR and has no transcoded file.

```
docker logs -f immich_server 2>&1 | grep -i "encod\|transcode\|skip"
Transcoded video exists for asset xxxx, but is no longer required. Deleting...
```

The changes are lost if you "down" and "up" the stack, so I did another ugly workaround to have it persistent.

Create a directory where you keep your docker-compose.yml (can be actually anywhere, but just for keeping things sane)

```
mkdir -p /path/to/immich-app/patches
```

Copy the files

```
docker cp immich_server:/usr/src/app/server/dist/utils/media.js /path/to/immich-app/patches/media.js
docker cp immich_server:/usr/src/app/server/dist/services/media.service.js /path/to/immich-app/patches/media.service.js
```

Modify docker-compose.yml and add the following under volumes: in the immich-server: section

      - /path/to/immich-app/patches/media.js:/usr/src/app/server/dist/utils/media.js:ro
      - /path/to/immich-app/patches/media.service.js:/usr/src/app/server/dist/services/media.service.js:ro
      
 Now you can restart the stack and have persistent changes. 
 
 Good luck, I hope it works also for you if you want to try!
